### PR TITLE
#1 - Add initial commit

### PR DESCRIPTION
**Pull request by ebenezer-melkamu**
Fixes #1

### Changelog 
 1. Added UI and API to Create, List, and Edit Genres 

### Issues 
1. React admin requires an id key to show the data, but the soul API returns `GenereId` instead of an  `id`  property, so after i get the data from the Soul API, i am running a loop in the dataProvider to add an `id` key 
2. For now i couldn't display the Edit button in the Datagrid, so you have to click the item to edit it (i will fix this issue later)
3. For now the `getList` providers only work for the `/tables/genre/rows` API

### CODE REVIEW

**Review by IanMayo — CHANGES_REQUESTED** [review]
One change, please.  The data-provider maps `GenreId` to `id`.  We will have to do this for each resource type. I think a `good engineering` approach to this would be to have a dictionary of primary ids for resources.  So, for `Genres/rows` we would have `GenreId`.   Then at line `26` of `dataProvider.js` you would look up the name of the primary key to use.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,67 @@
-# Soul4ReactAdmin
-Demonstrator for using Soul database wrapper with ReactAdmin
+# Soul for React Admin
```
Comment (IanMayo) [inline README.md:1]: This is a good quality piece of work. Thanks @AbegaM /

File: client/src/components/Genre.jsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  Datagrid,
+  TextField,
+  Create,
+  SimpleForm,
+  TextInput,
+  Edit,
+  useRecordContext,
+} from "react-admin";
+
+function GenreTitle() {
```
Comment (IanMayo) [inline client/src/components/Genre.jsx]: this function doesn't get used

Comment (ebenezer-melkamu) [reply client/src/components/Genre.jsx]: yes, i was about to use that to show the title, i can remove it

**Comment by ebenezer-melkamu** [comment]
> One change, please. The data-provider maps `GenreId` to `id`. We will have to do this for each resource type. I think a `good engineering` approach to this would be to have a dictionary of primary ids for resources. So, for `Genres/rows` we would have `GenreId`. Then at line `26` of `dataProvider.js` you would look up the name of the primary key to use.

ok no problem, i will add this feature

**Comment by ebenezer-melkamu** [comment]
@IanMayo I have added the primary key dictionary in the dataProvider, let me know what you think about it.

**Comment by IanMayo** [comment]
Hello @AbegaM - I have a general comment on your approach. We're creating a DataProvider that is specific to `soul`.

So, I think it would be ok to treat the `rows` part of the URL as a soul-specific detail.  So, in `app.js`, we would declare the resources as `Genres`, not `Genres/rows`.  We would insert the `rows` element in the `DataProvider`.  In this way, our `app.js` is ignorant of `soul` specifics - and it's easier to switch to another data-provider.

So, the change would be in `app.js` and `dataProvider.js`.

File: client/src/common/dataProvider.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
+import { stringify } from "query-string";
+import axios from "axios";
+
+import config from "../config";
+
+const pkDictionary = {
```
Comment (IanMayo) [inline client/src/common/dataProvider.js]: I'd like to request a change here, please.  Could this dictionary be passed into the `dataProvider`?  Then the `dataProvider` doesn't need to know about our specific `chinook.db` database.

Also, when you're doing the `id` swap, you would use something like:
```
cont primaryKey = pkDictionary[resource]
if (primaryKey !== undefined) {
  item.id = item[primaryKey]
}
```

That would avoid having to declare the primary keys for a database that uses plain `id` fields.

**Comment by IanMayo** [comment]
Aah, @AbegaM - I see you still have some more verbs to support, too - including `delete`.  It may be a challenge to create `react-admin` UI elements that drive the full API, but I'll try and support you in that.

File: client/src/common/dataProvider.js (hunk 3 above)
Comment (ebenezer-melkamu) [reply client/src/common/dataProvider.js]: Hello Ian, can you please give me some explanation on this comment, currently the `pkDictionary` is in the `dataProvider.js` file, do you want me to add the `pkDictionary` as an object like this and then access it in the `getList` method 

```js 
export const dataProvider = {
  getList: (resource, params) => {}, 
  getOne: (resource, params) => {}, 
  configs: {pkDictionary: {} }
}
```

**Comment by ebenezer-melkamu** [comment]
> Aah, @AbegaM - I see you still have some more verbs to support, too - including `delete`. It may be a challenge to create `react-admin` UI elements that drive the full API, but I'll try and support you in that.

No problem, is there any other verb i have to add otherthan `create`, `list`, `edit`, and `delete`? let me know if there are any and i will add them

File: client/src/common/dataProvider.js (hunk 3 above)
Comment (IanMayo) [reply client/src/common/dataProvider.js]: Sure, I'd like to remove `dataProvider`s knowledge of the structure of the chinook database.  So, the primary keys would be defined in `app.js`, and passed in as a parameter of `dataProvider`.  I'm pretty sure other react-admin data-providers take parameters, particularly for the URL to use.  Aah, the URL should be passed in too, please.   Then, no changes need to be made to `dataProvider` in order to support a different `soul` database.

```
  export const dataProvider = (pkDictionary) => {
  getList: (resource, params) => {
    const primaryKey = pkDictionary[resource]
}, 
  getOne: (resource, params) => {}, 
  configs: {pkDictionary: {} }
}
```

**Comment by ebenezer-melkamu** [comment]
@IanMayo i have made two changes, please give me your comments 
1. I tried to find how to pass parameters to the dataProvider in the documentation, but i haven't found one, so i am calling the `dataProvider()` as a function in `app.js` and i am passing the `pkDictionary` and the `url` from there. 
2. I have added a `delete` dataProvider, but i couldn't delete elements using the `soul` api and the reason is that `soul` keeps returning `SQLIte error` 
 ```
    SQLITE_CONSTRAINT_FOREIGNKEY 
    FOREIGN KEY constraint failed
 ``` 
I think the reason for this error is that, i am trying to delete a `Genre` row and i am passing the `GenreId` in the URL but this `GenreId` exists in other tables as a foreign key.

**Comment by IanMayo** [comment]
How about loading another table - once that isn't the foreign key for other data?

(I don't have database here with me, so I can't tell you which one)

**Comment by ebenezer-melkamu** [comment]
> How about loading another table - once that isn't the foreign key for other data?
> 
> (I don't have database here with me, so I can't tell you which one)

ok, let me try that

**Comment by IanMayo** [comment]
> 1. I tried to find how to pass parameters to the dataProvider in the documentation, but i haven't found one, so i am calling the `dataProvider()` as a function in `app.js` and i am passing the `pkDictionary` and the `url` from there.

There is an example here of passing a parameter to a data provider:
https://marmelab.com/react-admin/DataProviders.html

The data provider itself is here:
https://github.com/marmelab/react-admin/blob/master/packages/ra-data-simple-rest/src/index.ts

Aah, but, err... `react-admin` have moved to TypeScript, but we're not (yet) using TypeScript in our project.  I'll try to make the change for you tomorrow morning.

**Comment by ebenezer-melkamu** [comment]
ok no problem, i also can do the typescript migration

**Comment by IanMayo** [comment]
> ok no problem, i also can do the typescript migration

Ok, but let's defer that for another PR #3

File: client/src/common/dataProvider.js [hunk 4]
```diff
@@ -1,76 +1,77 @@
...
+        return { data: { id: response.data.lastInsertRowId, ...params.data } };
+      });
+    },
+
+    getOne: (resource, params) => {
+      const url = `${apiUrl}/${resource}/rows/${params.id}`;
+
+      return axios.get(url).then((response) => {
+        const { data } = response.data;
+        return { data: { id: data[0].GenreId, ...data[0] } };
```
Comment (IanMayo) [inline client/src/common/dataProvider.js]: This is only valid for the `Genres` table.   We have to do the `pk` test, and replace the id if there is a pk for this resource.  (so, remove the pl and insert the `id`.

File: client/src/common/dataProvider.js [hunk 5]
```diff
@@ -1,76 +1,77 @@
...
+      return axios.put(url, { fields: editData }).then((response) => {
+        return { data: { id: response.data.lastInsertRowId, ...params.data } };
+      });
+    },
+
+    deleteMany: (resource, params) => {
+      const ids = params.ids.toString();
+      const url = `${apiUrl}/${resource}/rows/${ids}`;
+
+      return axios.delete(url).then((response) => {
```
Comment (IanMayo) [inline client/src/common/dataProvider.js]: `react-admin` says the response should include the ids of the deleted records:
image

File: README.md [hunk 6]
```diff
@@ -1,2 +1,67 @@
...
+
+These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.
+
+## Running the soul wrapper
+
+1. Copy the `chinook.db` database file to the core folder
+
+   ```
+    cp chinook.db /soul/core
```
Comment (IanMayo) [inline README.md]: is there really a leading slash at the start of `/soul`?

**Comment by IanMayo** [comment]
@AbegaM - I'm not sure how I get the `soul` code into the application.

I haven't worked with submodules before. Could you update the `readme` to include the necessary step(s) please?

File: README.md (hunk 6 above)
Comment (ebenezer-melkamu) [reply README.md]: My bad, you don't need it, i will update the Readme

**Comment by ebenezer-melkamu** [comment]
> @AbegaM - I'm not sure how I get the `soul` code into the application.
> 
> I haven't worked with submodules before. Could you update the `readme` to include the necessary step(s) please?

ok, i will update the Readme

**Comment by ebenezer-melkamu** [comment]
> 

@IanMayo I have fixed the submodule issue, and i have updated the Readme, all you need to do is run the command below to clone the `main-repo` and the `submodule-repo` at the same time.
  ```
    git clone --recurse-submodules git@github.com:DeepBlueCLtd/Soul4ReactAdmin.git
  ```

**Comment by ebenezer-melkamu** [comment]
@IanMayo For the delete feature to work in `soul` i think i have to do the following 

Example 
* Lets say i want to delete this `Genre`  from the database 
   ```json
      {
        "GenreId": 1,
        "Name": "Rock"
      },
   ```
* This `Genre` has a relationship with the `Tracks` table, the `GenreId` is used as a foreign key in the `Track` table, so before deleting the `Genre` i have to delete all of the `Tracks` which have a `GenreId` of `1`.  There are around `1731` tracks that have a `GenreId` of `1` so when a user tries to delete this `Genre` from React-Admin, I will pull all of the `Tracks` from the Database and i will delete all of the tracks, and then i will delete the `Genre`, is this approach correct?

**Comment by IanMayo** [comment]
No, just switch to a table that isn't acting as a foreign key to another table.  It would be useful to extend your example to more than one `resource`  anyway.

File: client/src/common/dataProvider.js [hunk 7]
```diff
@@ -1,77 +1,106 @@
...
+    return axios.get(url).then((response) => {
+      const { data } = response.data;
+
+      //manually add an id key
+      const modifiedData = [];
+      for (let i = 0; i < data.length; i++) {
+        const item = data[i];
+        const primaryKey = pkDictionary[resource];
+
+        if (primaryKey !== undefined) {
```
Comment (IanMayo) [inline client/src/common/dataProvider.js:81]: @AbegaM - when we insert the new `id`, could remove the existing one please?  Having the unused id results in `ListGuesser` creating an extra column for it.  

That relates to all of the `get` actions, please.

Comment (ebenezer-melkamu) [reply client/src/common/dataProvider.js:81]: ok, i will fix this issue

**Comment by ebenezer-melkamu** [comment]
@IanMayo  
All providers are working except for the `updateMany` provider and the `getManyReferences` provider, i am not sure how to trigger this methods but i am working on it, the `delete` feature is working for the `Invoice Items` and all of the other providers are working for the `Genre` page.

**Comment by ebenezer-melkamu** [comment]
@IanMayo please let me know if you have any comments, i have pushed the search feature and i have fixed the above comments.

File: client/src/common/dataProvider.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+          item.id = item[primaryKey];
+          delete item[primaryKey];
+        }
+
+        modifiedData.push(item);
+      }
+
```
Comment (IanMayo) [inline client/src/common/dataProvider.js:39]: let me try a trendy JS-array way or writing the above block:
```
const primaryKey = pkDictionary[resource];
const modifiedData = primaryKey ? data.map((item) => {
  item.id = item[primaryKey];
  delete item[primaryKey];
  return item
}) : data
```

**Comment by IanMayo** [comment]
Nearly there, but I had CORS issues:
image

I do have the correct whitelist:
image

**Comment by ebenezer-melkamu** [comment]
> Nearly there, but I had CORS issues: image
> 
> I do have the correct whitelist: image

@IanMayo i think you don't need the `/` at the end 
  ```
   CORS_ORIGIN_WHITELIST=http://localhost:5173
  ```

**Comment by IanMayo** [comment]
@AbegaM - thanks, that fixed it!

**Comment by ebenezer-melkamu** [comment]
@IanMayo i don't see a big difference but here is mine
Screen Shot 2022-12-02 at 8 21 30 PM

**Comment by IanMayo** [comment]
Ok, I have app running.  "Edit" doesn't do anything for "invoice items" or "tracks".  I _Guess_ I need to edit `invoice items` to test `delete` functionality.

**Comment by ebenezer-melkamu** [comment]
> Ok, I have app running. "Edit" doesn't do anything for "invoice items" or "tracks". I _Guess_ I need to edit `invoice items` to test `delete` functionality.

ohh, i have messed up something in my last commit, i will fix it

**Comment by ebenezer-melkamu** [comment]
@IanMayo I have fixed the edit feature, but you can only delete `invoice_items`, the `Tracks` table is returning a `SQLite foreignKey error` like the `Genere` table

**Review by IanMayo — APPROVED** [review]
Working!

**Pull request merged**